src/extractor/extraction_containers.cpp

```diff
@@ -1159,6 +1210,95 @@ ExtractionContainers::ReferencedWays ExtractionContainers::IdentifyRestrictionWa
+
+    // Temporary store for nodes containing a unidirectional signal.
+    std::unordered_map<OSMNodeID, StopSign::Direction> unidirectional_signals;
+
+    // For each node that has a unidirectional traffic signal, we store the node(s)
+    // that lead up to the signal.
+    std::unordered_multimap<OSMNodeID, OSMNodeID> signal_segments;
+
+    std::unordered_set<OSMNodeID> bidirectional_signals;
+
+    const auto mark_signals = [&](auto const &traffic_signal) {
+        if (traffic_signal.second == StopSign::DIRECTION_FORWARD ||
+            traffic_signal.second == StopSign::DIRECTION_REVERSE)
+        {
+            unidirectional_signals.insert({traffic_signal.first, traffic_signal.second});
+        }
+        else
+        {
+            BOOST_ASSERT(traffic_signal.second == StopSign::DIRECTION_ALL);
+            bidirectional_signals.insert(traffic_signal.first);
+        }
+    };
+    std::for_each(external_stop_signs.begin(), external_stop_signs.end(), mark_signals);
+
+    // Extract all the segments that lead up to unidirectional traffic signals.
+    const auto set_segments = [&](const size_t way_list_idx, auto const & /*unused*/) {
+        const auto node_start_offset =
+            used_node_id_list.begin() + way_node_id_offsets[way_list_idx];
+        const auto node_end_offset =
+            used_node_id_list.begin() + way_node_id_offsets[way_list_idx + 1];
+
+        for (auto node_it = node_start_offset; node_it < node_end_offset; node_it++)
+        {
+            const auto sig = unidirectional_signals.find(*node_it);
+            if (sig != unidirectional_signals.end())
+            {
+                if (sig->second == StopSign::DIRECTION_FORWARD)
+                {
+                    if (node_it != node_start_offset)
+                    {
+                        // Previous node leads to signal
+                        signal_segments.insert({*node_it, *(node_it - 1)});
+                    }
+                }
+                else
+                {
+                    BOOST_ASSERT(sig->second == StopSign::DIRECTION_REVERSE);
+                    if (node_it + 1 != node_end_offset)
+                    {
+                        // Next node leads to signal
+                        signal_segments.insert({*node_it, *(node_it + 1)});
+                    }
+                }
+            }
+        }
+    };
+    util::for_each_indexed(ways_list.cbegin(), ways_list.cend(), set_segments);
+
+    util::for_each_pair(
+        signal_segments, [](const auto pair_a, const auto pair_b) {
+            if (pair_a.first == pair_b.first)
+            {
+                // If a node is appearing multiple times in this map, then it's ambiguous.
+                // The node is an intersection and the traffic direction is being use for multiple
+                // ways. We can't be certain of the original intent. See:
+                // https://wiki.openstreetmap.org/wiki/Key:traffic_signals:direction
```

Comment:
Might not be ambiguous for stop signs, would have to see what the OSM consensus is.